src/dbnode/storage/index.go

```diff
@@ -2098,13 +2095,90 @@ func (i *nsIndex) CleanupExpiredFileSets(t time.Time) error {
 	return i.deleteFilesFn(filesets)
 }
 
-func (i *nsIndex) CleanupDuplicateFileSets() error {
+type corruptedInfoFile struct {
+	IndexVolumeType idxpersist.IndexVolumeType
+	File            fs.ReadIndexInfoFileResult
+}
+
+func (i *nsIndex) CleanupCorruptedFileSets() error {
+	/*
+	   Corrupted index filesets can be safely cleaned up if its not
+	   the latest volume index per index volume type/block start combo.
+
+	   We are guaranteed not to be actively writing to an index fileset once
+	   we're already writing to later volume indices.
+	*/
 	fsOpts := i.opts.CommitLogOptions().FilesystemOptions()
-	infoFiles := i.readIndexInfoFilesFn(
-		fsOpts.FilePathPrefix(),
-		i.nsMetadata.ID(),
-		fsOpts.InfoReaderBufferSize(),
+	infoFiles := i.readIndexInfoFilesFn(fs.ReadIndexInfoFilesOptions{
+		FilePathPrefix:   fsOpts.FilePathPrefix(),
+		Namespace:        i.nsMetadata.ID(),
+		ReaderBufferSize: fsOpts.InfoReaderBufferSize(),
+		IncludeCorrupted: true,
+	})
+
+	var (
+		toDelete                      []string
+		corrupted                     []corruptedInfoFile
+		latestBlockStart              int64
+		latestVolumeIndexByVolumeType = make(map[idxpersist.IndexVolumeType]int)
 	)
+	// NB: Info files should be ordered by block start.
+	for _, file := range infoFiles {
+		// This intentionally skips the latest block start as that's the active block.
+		if file.Info.BlockStart > latestBlockStart {
+			for volType := range latestVolumeIndexByVolumeType {
+				latestVolumeIndex := latestVolumeIndexByVolumeType[volType]
+				// Reset latest vol idx.
+				latestVolumeIndexByVolumeType[volType] = 0
+
+				// NB: We only remove stale index filesets to avoid deleting
+				// one we're actively writing to.
+				for _, cf := range corrupted {
+					if cf.IndexVolumeType != volType {
+						continue
+					}
+					if cf.File.ID.VolumeIndex >= latestVolumeIndex {
+						continue
+					}
+					toDelete = append(toDelete, cf.File.AbsoluteFilePaths...)
+				}
+			}
+			latestBlockStart = file.Info.BlockStart
+			corrupted = corrupted[:0]
+		}
+
+		// NB: Missing info file data means the info file itself is corrupt.
+		// When we start writing an index volume, we first write an incomplete index info file so
+		// we are always guaranteed its existence but not if it was written successfully.
+		if file.Info.BlockStart == 0 {
+			// Mark filesets w/ corrupted index info files for deletion right away.
+			toDelete = append(toDelete, file.AbsoluteFilePaths...)
+			continue
+		}
```

Comment:
I wonder if it is possible that the info file will be read after it was created, but before the contents were written? Not sure about filesystem atomicity guarantees, but at least `WriteFile()` implementation [[1](https://golang.org/src/io/ioutil/ioutil.go?s=1537:1605#L35), [2](https://golang.org/src/os/file.go?s=21309:21370#L685)] hints that there might be a gap. 

A comment above says that we intentionally skip latest block start. Could we accidentally violate this by reading info file which haven't been written yet and assuming it's corrupted? If that file belonged to the latest block start, we wouldn't even know.

Reply:
Hmm yea, I'm not sure - maybe @robskillington can comment on this. 

We _could_ err on the side of caution and not attempt to delete filesets w/ a potentially corrupt index info file.